src/jsdoc_transformer.ts

```diff
@@ -0,0 +1,937 @@
+  // Gather parameter properties from the constructor, if it exists.
+  const ctors: ts.ConstructorDeclaration[] = [];
+  let paramProps: ts.ParameterDeclaration[] = [];
+  const nonStaticProps: Array<ts.PropertyDeclaration|ts.PropertySignature> = [];
+  const staticProps: Array<ts.PropertyDeclaration|ts.PropertySignature> = [];
+  const unhandled: ts.NamedDeclaration[] = [];
+  const abstractMethods: ts.FunctionLikeDeclaration[] = [];
+  for (const member of typeDecl.members) {
+    if (member.kind === ts.SyntaxKind.Constructor) {
+      ctors.push(member as ts.ConstructorDeclaration);
+    } else if (ts.isPropertyDeclaration(member) || ts.isPropertySignature(member)) {
+      const isStatic = transformerUtil.hasModifierFlag(member, ts.ModifierFlags.Static);
+      if (isStatic) {
+        staticProps.push(member);
+      } else {
+        nonStaticProps.push(member);
+      }
+    } else if (
+        member.kind === ts.SyntaxKind.MethodDeclaration ||
+        member.kind === ts.SyntaxKind.MethodSignature ||
+        member.kind === ts.SyntaxKind.GetAccessor || member.kind === ts.SyntaxKind.SetAccessor) {
+      if (transformerUtil.hasModifierFlag(member, ts.ModifierFlags.Abstract) ||
+          ts.isInterfaceDeclaration(typeDecl)) {
+        abstractMethods.push(
+            member as ts.MethodDeclaration | ts.GetAccessorDeclaration | ts.SetAccessorDeclaration);
+      }
+      // Non-abstract methods only exist on classes, and are handled in regular emit.
+    } else {
+      unhandled.push(member);
+    }
+  }
+
+  if (ctors.length > 0) {
+    const ctor = ctors[0];
+    paramProps = ctor.parameters.filter(
+        p => transformerUtil.hasModifierFlag(p, FIELD_DECLARATION_MODIFIERS));
+  }
+
+  if (nonStaticProps.length === 0 && paramProps.length === 0 && staticProps.length === 0 &&
+      abstractMethods.length === 0) {
+    // There are no members so we don't need to emit any type
+    // annotations helper.
+    return null;
+  }
+
+  if (!typeDecl.name) {
+    mtt.debugWarn(typeDecl, 'cannot add types on unnamed declarations');
+    return null;
+  }
+
+  const className = transformerUtil.getIdentifierText(typeDecl.name);
+  const staticPropAccess = ts.createIdentifier(className);
+  const instancePropAccess = ts.createPropertyAccess(staticPropAccess, 'prototype');
+  // Closure Compiler will report conformance errors about this being unknown type when emitting
+  // class properties as {?|undefined}, instead of just {?}. So make sure to only emit {?|undefined}
+  // on interfaces.
+  const isInterface = ts.isInterfaceDeclaration(typeDecl);
+  const propertyDecls = [
+    ...staticProps.map(
+        p => createClosurePropertyDeclaration(
+            mtt, staticPropAccess, p, isInterface && !!p.questionToken)),
+    ...nonStaticProps.map(
+        (p) => createClosurePropertyDeclaration(
+            mtt, instancePropAccess, p, isInterface && !!p.questionToken)),
+    ...paramProps.map(
```

Comment:
Does it work to do
[...staticProps, ...nonStaticProps, ...paramProps].map((p) => ...

Reply:
Nice, done.